File: app/src/main/java/br/pedroso/upcomingmovies/movieslist/MoviesListPresenter.java
package br.pedroso.upcomingmovies.movieslist;

import android.util.Log;

import java.util.List;

import javax.inject.Inject;

import br.pedroso.upcomingmovies.domain.Movie;
import br.pedroso.upcomingmovies.domain.MoviesRepository;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.disposables.Disposable;

public class MoviesListPresenter {
    private static final String LOG_TAG = MoviesListPresenter.class.getName();

    private final MoviesListView view;

    private final MoviesRepository moviesRepository;

    private final CompositeDisposable disposables = new CompositeDisposable();

    @Inject
    MoviesListPresenter(MoviesListView view, MoviesRepository moviesRepository) {
        this.moviesRepository = moviesRepository;
        this.view = view;
    }

    public void resume() {
        loadUpcomingMovies();
    }

    private void loadUpcomingMovies() {
        view.cleanMoviesList();

        Disposable disposable = moviesRepository.listUpcomingMovies()
                .observeOn(AndroidSchedulers.mainThread()) // TODO: improve this.
                .subscribe(this::displayLoadedMovies, this::displayErrorOnLoadingMessage);

        disposables.add(disposable);
    }

    private void displayErrorOnLoadingMessage(Throwable throwable) {
        // TODO
    }

    private void displayLoadedMovies(List<Movie> movies) {
        view.renderMoviesList(movies);
    }

    public void onMovieClick(Movie movie) {
        Log.d(LOG_TAG, "Clicked on the movie: " + movie);

        view.startMovieDetailsActivity(movie.getId());
    }

    public void pause() {
        disposables.clear();
    }
}

File: app/src/main/kotlin/br/pedroso/upcomingmovies/movieslist/MoviesListPresenter.kt
package br.pedroso.upcomingmovies.movieslist

import android.util.Log
import br.pedroso.upcomingmovies.domain.Movie
import br.pedroso.upcomingmovies.domain.MoviesRepository
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.disposables.CompositeDisposable
import javax.inject.Inject

class MoviesListPresenter @Inject internal constructor(
    private val view: MoviesListView,
    private val moviesRepository: MoviesRepository
) {
    private val disposables = CompositeDisposable()

    fun resume() {
        loadUpcomingMovies()
    }

    private fun loadUpcomingMovies() {
        view.cleanMoviesList()

        val disposable = moviesRepository.listUpcomingMovies()
            .observeOn(AndroidSchedulers.mainThread()) // TODO: improve this.
            .subscribe(
                { movies: List<Movie> -> this.displayLoadedMovies(movies) },
                { throwable: Throwable -> this.displayErrorOnLoadingMessage(throwable) })

        disposables.add(disposable)
    }

    private fun displayErrorOnLoadingMessage(throwable: Throwable) {
        Log.e(MoviesListPresenter::class.simpleName, "Failed to load movies.", throwable)
    }

    private fun displayLoadedMovies(movies: List<Movie>) {
        view.renderMoviesList(movies)
    }

    fun onMovieClick(movie: Movie) {
        Log.d(LOG_TAG, "Clicked on the movie: $movie")

        view.startMovieDetailsActivity(movie.id)
    }

    fun pause() {
        disposables.clear()
    }

    companion object {
        private val LOG_TAG: String = MoviesListPresenter::class.java.name
    }
}
